c10/core/impl/LocalDispatchKeySet.h

```diff
@@ -95,28 +85,85 @@ class C10_API IncludeDispatchKeyGuard {
+class ExcludeDispatchKeyGuard {
  public:
-  ExcludeDispatchKeyGuard(DispatchKeySet);
-  ExcludeDispatchKeyGuard(DispatchKey k)
-      : ExcludeDispatchKeyGuard(DispatchKeySet(k)) {}
+  // If our exclude set does not overlap with c10::default_excluded_set, we can
+  // skip some bookkeeping. (And we know at compile time if this is the case.)
+  // Key exclusion tends to be on the hot path, so it's worth it to bypass the
+  // unnecessary XORs if possible. We force `has_overlap` to be explicitly
+  // declared so that changes to the default excluded set do not silently knock
+  // us off the hot path.
+  static_assert(
+      has_overlap == (bool)(exclude & c10::default_excluded_set.raw_repr()),
+      "Declared `has_overlap` does not match computed value.");
+
   ExcludeDispatchKeyGuard(const ExcludeDispatchKeyGuard&) = delete;
   ExcludeDispatchKeyGuard operator=(const ExcludeDispatchKeyGuard&) = delete;
   ExcludeDispatchKeyGuard(ExcludeDispatchKeyGuard&&) = delete;
   ExcludeDispatchKeyGuard operator=(ExcludeDispatchKeyGuard&&) = delete;
-  ~ExcludeDispatchKeyGuard();
+
+  explicit ExcludeDispatchKeyGuard(PODLocalState* tls) : tls_(tls) {
+    if (has_overlap) {
+      LocalDispatchKeySetWrapper wrapper{tls_};
+      auto current_excluded = wrapper.excluded();
+      auto exclude_set = DispatchKeySet(DispatchKeySet::RAW, exclude);
+      delta_ = (exclude_set - current_excluded).raw_repr();
+      wrapper.set_excluded(current_excluded | exclude_set);
+
+    } else {
+      // Fast path
+      delta_ = exclude & ~(tls_->excluded_);
+      tls_->excluded_ |= exclude;
+    }
+  }
+
+  // Certain key guards (such as `AutoDispatchBelowADInplaceOrView`) are part
+  // of the public API, so we have to allow instantiation without exposing
+  // the implementation detail of `_get_thread_local_state()`.
+  ExcludeDispatchKeyGuard()
+      : ExcludeDispatchKeyGuard(_get_thread_local_state()) {}
+
+  ~ExcludeDispatchKeyGuard() {
+    if (has_overlap) {
+      LocalDispatchKeySetWrapper wrapper{tls_};
+      auto current = wrapper.excluded();
+      auto delta = DispatchKeySet(DispatchKeySet::RAW, delta_);
+      wrapper.set_excluded(current - delta);
+
+    } else {
+      // Fast path
+      tls_->excluded_ &= ~delta_;
+    }
+  };
 
  private:
   // A little micro-optimization to save us from tls_get_addr call
   // on destruction
-  PODLocalDispatchKeySet* tls_;
-  DispatchKeySet exclude_;
+  PODLocalState* tls_;
+  uint64_t delta_;
+};
+
+template <DispatchKey k, bool has_overlap>
+class ExcludeSingleDispatchKeyGuard {
+  static constexpr auto k_set = DispatchKeySet(k);
+  ExcludeDispatchKeyGuard<k_set.raw_repr(), has_overlap> guard_;
 };
 
+// Create a guard which can be exported by subclassing a specialization
+// of ExcludeDispatchKeyGuard. Note that because it will be created in ATen,
+// we need TORCH_API rather than C10_API.
+#define SPECIALIZE_EXCLUDE_GUARD(guard_name, k, has_overlap)              \
```

Comment:
Macros in headers need to be appropriately namespaced; e.g., `TORCH_SPECIALIZE_EXCLUDE_GUARD`

Reply:
(or `C10_SPECIALIZE_EXCLUDE_GUARD`, which may be more appropriate given this is in c10/ folder)